File: frontends/dart/lib/adapter/runanywhere.dart
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.

import 'package:ffi/ffi.dart';

import '../src/ffi/bindings.dart';
import 'voice_session.dart';

/// Public entry point — mirror of RunAnywhere.swift / RunAnywhere.kt.
///
/// Usage:
/// ```dart
///   final session = RunAnywhere.solution(
///       SolutionConfig.voiceAgent(VoiceAgentConfig()));
///   await for (final e in session.run()) {
///     // ...
///   }
/// ```
class RunAnywhere {
  const RunAnywhere._();

  static VoiceSession solution(SolutionConfig config) =>
      VoiceSession.create(config);

  /// Dynamic plugin load — Android/macOS/Linux only. Returns true on
  /// success. No-op on iOS (static plugin mode) where this returns false.
  static bool loadPlugin(String libPath) {
    try {
      final b = RaCoreBindings.open();
      final cpath = libPath.toNativeUtf8();
      final rc = b.loadPluginRaw(cpath);
      calloc.free(cpath);
      return rc == raOk;
    } catch (_) {
      return false;
    }
  }

  /// Count of currently-registered engine plugins.
  static int get registeredPluginCount {
    try {
      return RaCoreBindings.open().pluginCountRaw();
    } catch (_) {
      return 0;
    }
  }
}

sealed class SolutionConfig {
  const SolutionConfig._();

  factory SolutionConfig.voiceAgent(VoiceAgentConfig config) =
      VoiceAgentSolution;
  factory SolutionConfig.rag(RAGConfig config) = RAGSolution;
  factory SolutionConfig.wakeWord(WakeWordConfig config) = WakeWordSolution;
}

class VoiceAgentSolution extends SolutionConfig {
  final VoiceAgentConfig config;
  const VoiceAgentSolution(this.config) : super._();
}

class RAGSolution extends SolutionConfig {
  final RAGConfig config;
  const RAGSolution(this.config) : super._();
}

class WakeWordSolution extends SolutionConfig {
  final WakeWordConfig config;
  const WakeWordSolution(this.config) : super._();
}

class VoiceAgentConfig {
  final String llm;
  final String stt;
  final String tts;
  final String vad;
  final int    sampleRateHz;
  final int    chunkMs;
  final bool   enableBargeIn;
  final bool   emitPartials;
  final bool   emitThoughts;
  final String systemPrompt;
  final int    maxContextTokens;
  final double temperature;

  const VoiceAgentConfig({
    this.llm             = 'qwen3-4b',
    this.stt             = 'whisper-base',
    this.tts             = 'kokoro',
    this.vad             = 'silero-v5',
    this.sampleRateHz    = 16000,
    this.chunkMs         = 20,
    this.enableBargeIn   = true,
    this.emitPartials    = true,
    this.emitThoughts    = false,
    this.systemPrompt    = '',
    this.maxContextTokens = 4096,
    this.temperature     = 0.7,
  });
}

class RAGConfig {
  final String embedModel;
  final String rerankModel;
  final String llm;
  final String vectorStorePath;
  final int    retrieveK;
  final int    rerankTop;

  const RAGConfig({
    this.embedModel      = 'bge-small-en-v1.5',
    this.rerankModel     = 'bge-reranker-v2-m3',
    this.llm             = 'qwen3-4b',
    this.vectorStorePath = '',
    this.retrieveK       = 24,
    this.rerankTop       = 6,
  });
}

class WakeWordConfig {
  final String model;
  final String keyword;
  final double threshold;
  final int    preRollMs;

  const WakeWordConfig({
    this.model     = 'kws-zipformer-gigaspeech',
    this.keyword   = 'hey mycroft',
    this.threshold = 0.5,
    this.preRollMs = 250,
  });
}

File: frontends/ts/src/adapter/RunAnywhere.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.
//
// RunAnywhere v2 — public TypeScript / React Native entry point.

import { VoiceSession } from './VoiceSession.js';

export type SolutionConfig =
  | { kind: 'voice-agent'; config: VoiceAgentConfig }
  | { kind: 'rag';         config: RAGConfig        }
  | { kind: 'wake-word';   config: WakeWordConfig   };

export interface VoiceAgentConfig {
  llm?:              string;
  stt?:              string;
  tts?:              string;
  vad?:              string;
  sampleRateHz?:     number;
  chunkMs?:          number;
  enableBargeIn?:    boolean;
  emitPartials?:     boolean;
  emitThoughts?:     boolean;
  systemPrompt?:     string;
  maxContextTokens?: number;
  temperature?:      number;
}

export interface RAGConfig {
  embedModel?:      string;
  rerankModel?:     string;
  llm?:             string;
  vectorStorePath?: string;
  retrieveK?:       number;
  rerankTop?:       number;
}

export interface WakeWordConfig {
  model?:     string;
  keyword?:   string;
  threshold?: number;
  preRollMs?: number;
}

export const RunAnywhere = {
  /** Open a VoiceAgent / RAG / WakeWord session from a solution config. */
  solution(config: SolutionConfig): VoiceSession {
    return VoiceSession.create(config);
  },

  /**
   * Dynamic plugin load — React Native / Node only. Delegates to the
   * registered NativePipelineBindings (the TurboModule / N-API wrapper)
   * when one is present; returns false otherwise.
   */
  loadPlugin(libPath: string): boolean {
    const hostAny = RunAnywhere as unknown as {
      _hostLoadPlugin?: (path: string) => boolean;
    };
    return hostAny._hostLoadPlugin
      ? hostAny._hostLoadPlugin(libPath)
      : false;
  },

  /**
   * Host plug-in (React Native / Node) installs its loader here at
   * startup. See VoiceSession.setNativeBindings for the analogous
   * per-session surface.
   */
  setHostLoadPlugin(fn: ((path: string) => boolean) | null): void {
    (RunAnywhere as unknown as { _hostLoadPlugin?: ((p: string) => boolean) | null })
      ._hostLoadPlugin = fn ?? undefined;
  },
};
